include/device.hpp

```diff
@@ -0,0 +1,574 @@
+
+#define GRAPH_BLER 1
+#define GRAPH_E11 (1 << 1)
+#define GRAPH_E21 (1 << 2)
+#define GRAPH_E31 (1 << 3)
+#define GRAPH_E12 (1 << 4)
+#define GRAPH_E22 (1 << 5)
+#define GRAPH_E32 (1 << 6)
+
+#define GRAPH_PIE (1 << 1)
+#define GRAPH_PI8 (1 << 2)
+#define GRAPH_PIF (1 << 3)
+#define GRAPH_POE (1 << 4)
+#define GRAPH_PO8 (1 << 5)
+#define GRAPH_POF (1 << 6)
+
+#define GRAPH_LDC (1 << 1)
+#define GRAPH_BIS (1 << 4)
+
+#define GRAPH_UNCR (1 << 7)
+
+#define FEATURE_POWEREC 0x00000001
+#define FEATURE_HIDECDR 0x00000002
+#define FEATURE_SINGLESESSION 0x00000004
+#define FEATURE_SPEEDREAD 0x00000008
+#define FEATURE_BITSETR 0x00000010
+#define FEATURE_BITSETRDL 0x00000020
+#define FEATURE_SIMULPLUS 0x00000040
+
+#define FEATURE_GIGAREC 0x00000080
+#define FEATURE_VARIREC_CDBASE 0x00000100
+#define FEATURE_VARIREC_CDEXT 0x00000200
+#define FEATURE_VARIREC_CD (FEATURE_VARIREC_CDBASE | FEATURE_VARIREC_CDEXT)
+#define FEATURE_VARIREC_DVD 0x00000400
+#define FEATURE_VARIREC (FEATURE_VARIREC_CD | FEATURE_VARIREC_DVD)
+#define FEATURE_SECUREC 0x00000800
+#define FEATURE_SILENT 0x00001000
+
+#define FEATURE_AS_BASE 0x00002000
+#define FEATURE_AS_EXT 0x00004000
+#define FEATURE_AS (FEATURE_AS_BASE | FEATURE_AS_EXT)
+#define FEATURE_DESTRUCT 0x00008000
+#define FEATURE_F1TATTOO 0x00010000
+#define FEATURE_PIOLIMIT 0x00020000
+#define FEATURE_PIOQUIET 0x00040000
+
+#define FEATURE_LOEJ 0x00080000
+#define FEATURE_LOEJ_TOGGLE 0x00100000
+#define FEATURE_LOCK 0x00200000
+#define FEATURE_LOCK_TOGGLE 0x00400000
+
+#define AS_ACTION_MODE 1
+#define AS_ACTION_ACT 2
+#define AS_ACTION_DEACT 3
+#define AS_ACTION_DEL 4
+#define AS_ACTION_CLEAR 5
+#define AS_ACTION_CRE 6
+#define AS_ACTION_MQCK 7
+
+#define AS_MODE_OFF 0
+#define AS_MODE_AUTO 1
+#define AS_MODE_ON 2
+#define AS_MODE_FORCED 3
+
+#define ASCRE_FULL 8
+#define ASCRE_REPLACE 16
+
+#define ASMQCK_ADV 32
+
+#define PIOQ_QUIET 0
+#define PIOQ_STD 1
+#define PIOQ_PERF 2
+
+#define QPX_FILE_SIGN "QPXD"
+
+#if 0
+// chunk. cnout = 1
+#define META_TEST_ERRC_SUMMARY 57
+#define META_TEST_JB_SUMMARY 58
+#define META_TEST_FT_SUMMARY 59
+#endif
+
+struct ASDB_item {
+  bool present, active;
+  QString type;
+  QString mid, speed, writes;
+};
+
+typedef QList<ASDB_item> ASDB;
+
+struct DevFeatures {
+  int supported;
+  int enabled;
+
+  int prec_spd;
+  float grec;
+  char vrec_cd_pwr;
+  int vrec_cd_str;
+  char vrec_dvd_pwr;
+  int vrec_dvd_str;
+
+  QString sr_pass;
+
+  int sm_cd_rd, sm_cd_wr, sm_dvd_rd;
+  bool sm_access; // 0 = slow, 1 = fast
+  int sm_trayl, sm_traye;
+  bool sm_nosave;
+
+  int psm_cd_rd, psm_cd_wr, psm_dvd_rd;
+  //			psm_dvd_wr;
+  bool psm_access; // 0 = slow, 1 = fast
+  int psm_trayl, psm_traye;
+
+  int tattoo_inner, tattoo_outer;
+  QString tattoo_file;
+
+  int as_action;
+  int as_act_mode;
+  int as_mode;
+  int as_idx;
+  int as_mqckspd;
+  QString as_mqckres;
+  int pioq_quiet;
+  bool pioq_nosave;
+};
+
+struct DI_Transfer {
+  quint64 lba;
+  float spdx;
+  uint32_t spdk;
```

Comment:
Codacy found an issue: [struct member 'DI_Transfer::spdk' is never used.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)